server/routerlicious/packages/routerlicious-base/src/utils/sessionHelper.ts

```diff
@@ -0,0 +1,77 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { IDocumentSession, ISession } from "@fluidframework/server-services-client";
+import { IDocument, ICollection } from "@fluidframework/server-services-core";
+import { BaseTelemetryProperties, Lumberjack } from "@fluidframework/server-services-telemetry";
+
+export async function getSession(documentId: string,
+    ordererUrl: string,
+    historianUrl: string,
+    tenantId: string,
+    documentsCollection: ICollection<IDocument>): Promise<IDocumentSession> {
+    const lumberjackProperties = {
+        [BaseTelemetryProperties.tenantId]: tenantId,
+        [BaseTelemetryProperties.documentId]: documentId,
+    };
+
+    const tempDocument: IDocument = await documentsCollection.findOne({ documentId });
+    let tempSession: ISession = tempDocument.session;
+    const deli = JSON.parse(tempDocument.deli);
+    const isScribeEmpty = tempDocument.scribe === "";
+    const scribe = isScribeEmpty ? "" : JSON.parse(tempDocument.scribe);
+    let hasSessionLocationChanged: boolean = false;
+    const isSessionAlive: boolean = tempSession !== undefined ? tempSession.isSessionAlive : true;
+    if (tempSession === undefined) {
+        tempSession = {
+            ordererUrl,
+            historianUrl,
+            isSessionAlive: true,
+        };
+        await documentsCollection.upsert({
+            documentId,
+        }, {
+            deli: JSON.stringify(deli),
+            scribe: isScribeEmpty ? "" : JSON.stringify(scribe),
+            session: tempSession,
+        }, {
+        });
+        const ms: string = `The Session ${JSON.stringify(tempSession)} was inserted into the document collection`;
+        Lumberjack.info(ms, lumberjackProperties);
+    }
+
+    if (!tempSession.isSessionAlive) {
+        // Reset logOffset, ordererUrl, and historianUrl when switching cluster.
+        if (tempSession.ordererUrl !== ordererUrl) {
+            const ms: string = `Reset logOffset, ordererUrl, and historianUrl when switching cluster.`;
+            Lumberjack.info(ms, lumberjackProperties);
+            deli.logOffset = -1;
+            tempSession.ordererUrl = ordererUrl;
+            tempSession.historianUrl = historianUrl;
+            hasSessionLocationChanged = true;
+            if (!isScribeEmpty) {
+                scribe.logOffset = -1;
+            }
+        }
+        tempSession.isSessionAlive = true;
+        await documentsCollection.upsert({
+            documentId,
+        }, {
+            deli: JSON.stringify(deli),
+            scribe: isScribeEmpty ? "" : JSON.stringify(scribe),
+            session: tempSession,
+        }, {
+        });
+    }
+
+    tempSession.isSessionAlive = isSessionAlive;
```

Comment:
This overrides line 58 above. In line 58, we are doing `tempSession.isSessionAlive = true;` if `tempSession.ordererUrl !== ordererUrl`. Can that be a problem?

Let's say we have the following sequence of events:
1. In line 26, `tempSession.isSessionAlive` is `false`. So `isSessionAlive` also becomes `false`.
2. Because `tempSession.isSessionAlive` is `false`, we enter the if condition in line 45.
3. So, in line 58, we make `tempSession.isSessionAlive` be `true`.
4. `isSessionAlive` was never updated, so it is `false`. And when you get to line 69, you overwrite `tempSession.isSessionAlive` (which was made `true` in line 58), making it `false`.

Was that intentional?

Reply:
Yes. Assume we open a new session from a document that already closed, where the isSessionAlive flag is false in the DB. 

When opening it, we need to update the isSessionAlive as true and store it in the database. However, we return the actual isSessionAlive flag to the client, the client would know this document has the session closed when we open it.

Reply:
Can we add some comments documenting that decision in the code above? The explanation makes sense - you are saying that while `getSession()` will return to the caller with the status of the session, it needs to also update that status to signal that the session is now active (but the caller needs to know whether the session was alive before their request came).

However, it might be hard to infer all that from the code alone. I think some comments could help there.

Reply:
Yup. Added the logs to show the above workflow...